frontend/packages/console-shared/src/components/dynamic-form/utils.ts

```diff
@@ -54,30 +54,100 @@ export const hasNoFields = (jsonSchema: JSONSchema6 = {}, uiSchema: UiSchema = {
   const noUIFieldOrWidget = !uiSchema?.['ui:field'] && !uiSchema?.['ui:widget'];
   switch (type) {
-    case SchemaType.array:
+    case JSONSchemaType.array:
       return noUIFieldOrWidget && hasNoFields(jsonSchema.items as JSONSchema6, uiSchema?.items);
-    case SchemaType.object:
+    case JSONSchemaType.object:
       return (
         noUIFieldOrWidget &&
         _.every(jsonSchema?.properties, (property, propertyName) =>
           hasNoFields(property as JSONSchema6, uiSchema?.[propertyName]),
         )
       );
-    case SchemaType.boolean:
-    case SchemaType.integer:
-    case SchemaType.number:
-    case SchemaType.string:
+    case JSONSchemaType.boolean:
+    case JSONSchemaType.integer:
+    case JSONSchemaType.number:
+    case JSONSchemaType.string:
       return false;
     default:
       return noUIFieldOrWidget;
   }
 };
 
-// Given a JSONSchema and associated uiSchema, create the appropriat ui schema order property for
-// the jsonSchema. Orders properties according to the following rules:
+/**
+ * Give a property name a sort wieght based on whether it has ui schema, is required, or is a
+ * control field for a property with a field dependency. A lower weight means higher sort order.
+ * Fields are weighted according to the following tiers:
+ *  Tier 1 (negative 10^9 - 10^6 magnitude):  Required fields with ui schema
+ *  Tier 2 (negative 10^9 magnitude):         Required fields without ui schema
+ *  Tier 3 (negative 10^6 magnitude):         Optional fields with ui schema
+ *  Tier 4 (positive 10^3 maginitude):        Control fields that don't fit any above
+ *  Tier 5 (Infinity):                        All other fields
+ *
+ * Within each of the above tiers, fields are further weighted based on field dependency and ui
+ * schema defined sort order:
+ *   - Fields without dependency - base weight + ui schema sort order
+ *   - Control field - base weight + ui schema sort order + (nth control field) * 10000
+ *   - Dependent field - control field weight + ui schema sort order + 1
+ *
+ * These weight numbers are arbitrary, but spaced far enough apart to prevent collisions.
+ */
+const getJSONSchemaPropertySortWeight = (
+  property: string,
+  jsonSchema: JSONSchema6,
+  uiSchema: UiSchema,
+): number => {
+  const isRequired = (jsonSchema?.required ?? []).includes(property);
+  const propertyUISchema = uiSchema?.[property];
+
+  // Any sibling has a dependency with this as the control field.
+  const isControlField = _.some(
+    uiSchema,
+    ({ 'ui:dependency': dependency }) => dependency?.controlFieldName === property,
+  );
+
+  // Property sort order from ui schema (adjusted to 1-based origin). Use propertyNames.length
+  // when no uiSchema sortOrder exists. Ensures properties without uiSchema sort order  have a
+  // higher weight than those with.
+  const propertySortOrder = Number(
+    propertyUISchema?.['ui:sortOrder'] ?? _.keys(jsonSchema?.properties).length,
+  );
+
```

Comment:
This is the main bug fix. I added logic to include the original descriptor index as a sort weight modifier.